Brokerages/Binance/BinanceBrokerage.Messaging.cs

```diff
@@ -0,0 +1,317 @@
+                );
+
+                if (status == OrderStatus.Filled)
+                {
+                    Orders.Order outOrder;
+                    CachedOrderIDs.TryRemove(order.Id, out outOrder);
+                }
+
+                OnOrderEvent(orderEvent);
+            }
+            catch (Exception e)
+            {
+                Log.Error(e);
+                throw;
+            }
+        }
+
+        private Orders.Order FindOrderByExternalId(string brokerId)
+        {
+            var order = CachedOrderIDs
+                    .FirstOrDefault(o => o.Value.BrokerId.Contains(brokerId))
+                    .Value;
+            if (order == null)
+            {
+                order = _algorithm.Transactions.GetOrderByBrokerageId(brokerId);
+            }
+
+            return order;
+        }
+
+        private void FetchOrderBookSnapshot(Symbol symbol, BinanceOrderBook orderBook)
+        {
+            LockStream();
+
+            var endpoint = $"/api/v1/depth?symbol={_symbolMapper.GetBrokerageSymbol(symbol)}&limit=1000";
+            var request = new RestRequest(endpoint, Method.GET);
+            var response = ExecuteRestRequest(request);
+
+            if (response.StatusCode != HttpStatusCode.OK)
+            {
+                OnMessage(new BrokerageMessageEvent(BrokerageMessageType.Warning, -1, $"Can't fetch snapshot for symbol {symbol.Value}"));
+                return;
+            }
+
+            var snapshot = JsonConvert.DeserializeObject<Messages.OrderBookSnapshotMessage>(response.Content);
+
+            orderBook.BestBidAskUpdated -= OnBestBidAskUpdated;
+            orderBook.LastUpdateId = snapshot.LastUpdateId;
+            ProcessOrderBookEvents(orderBook, snapshot.Bids, snapshot.Asks);
+
+            EmitQuoteTick(
+                symbol,
+                orderBook.BestBidPrice,
+                orderBook.BestBidSize,
+                orderBook.BestAskPrice,
+                orderBook.BestAskSize);
+            orderBook.BestBidAskUpdated += OnBestBidAskUpdated;
+            UnlockStream();
+        }
+
+        private void ProcessOrderBookEvents(DefaultOrderBook orderBook, object[][] bids, object[][] asks)
+        {
+            foreach (var item in bids)
+            {
+                var price = (item[0] as string).ToDecimal();
+                var quantity = (item[1] as string).ToDecimal();
+                if (quantity == 0)
+                    orderBook.RemoveBidRow(price);
+                else
+                    orderBook.UpdateBidRow(price, quantity);
+            }
+
+            foreach (var item in asks)
+            {
+                var price = (item[0] as string).ToDecimal();
+                var quantity = (item[1] as string).ToDecimal();
+                if (quantity == 0)
+                    orderBook.RemoveAskRow(price);
+                else
+                    orderBook.UpdateAskRow(price, quantity);
```

Comment:
outside the scope of this PR, but I think for encapsulation purposes it makes sense for `UpdateAskRow` to perform the `quantity == 0` check internally to guarantee its own state is consistent.